cmd/uhc/cluster/login/login.go

```diff
@@ -121,11 +127,83 @@ func run(cmd *cobra.Command, argv []string) {
 	clusterResource := resource.Cluster(argv[0])
 
 	// Retrieve the cluster info:
-	response, err := clusterResource.Get().
-		Send()
+	response, err := clusterResource.Get().Send()
+
+	// If it's not an cluster id, try to query clusters using search param, we assume the
+	// list should be within 50, if not, we need update the cluster names prefix to minimize
+	// the list
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Can't retrieve clusters: %s\n", err)
-		os.Exit(1)
+		pageSize := 50
+		pageIndex := 1
+		listRequest := resource.List().
+			Size(pageSize).
+			Page(pageIndex)
+		listRequest.Search("name like '" + argv[0] + "'")
+		listResponse, err := listRequest.Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't retrieve clusters: %s\n", err)
+			os.Exit(1)
+		}
+		clusters := make(map[string][]string)
+		listResponse.Items().Each(func(cluster *v1.Cluster) bool {
+			api := cluster.API().URL()
+			clusterName := cluster.Name()
+			clusterid := cluster.ID()
+			clusters[clusterid] = []string{clusterName, api}
+			return true
+		})
+		if len(clusters) == 0 {
+			fmt.Fprint(os.Stderr, "Can't retrieve clusters. \n")
+			os.Exit(1)
+		}
+		if len(clusters) == 1 {
+			var clusterid, clusterName string
+			for k, v := range clusters {
+				clusterid = k
+				clusterName = v[0]
+			}
+			fmt.Printf("Only one cluster match the args, will login to cluster:\n Name: %s\n ID: %s\n", clusterName, clusterid)
+			response, err = resource.Cluster(clusterid).Get().Send()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Can't retrieve cluster: %v\n", err)
+				os.Exit(1)
+			}
+		} else {
+			clusterList := []string{}
+			for k, v := range clusters {
+				clusterList = append(clusterList, fmt.Sprintf("Name: %s, ID: %s", v[0], k))
+			}
+			choice := ""
+			prompt := &survey.Select{
+				Message: "Please choose a cluster:",
+				Options: clusterList,
+				Default: clusterList[0],
+			}
+			clusterid := ""
+			err = survey.AskOne(prompt, &choice, func(ans interface{}) error {
+				choice := ans.(string)
+				found := false
+				for k := range clusters {
+					if strings.Contains(choice, k) {
+						found = true
+						clusterid = k
+					}
+				}
+				if !found {
+					return fmt.Errorf("the cluster you choose is not valid: %s", choice)
+				}
+				return nil
+			})
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Failed to login to cluster: %v\n", err)
+				os.Exit(1)
+			}
+			response, err = resource.Cluster(clusterid).Get().Send()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Can't retrieve cluster: %v\n", err)
+				os.Exit(1)
+			}
+		}
```

Comment:
This section of code looks too long. Can you move the code that retrieves the list of potential clusters to a separate function? For example, you could have a function like this:

```go
// findClusters finds the clusters that match the given key. A cluster matches the key if its
// identifier is that key, or if its name starts with that key. For example, the key `prd-2305`
// doesn't match a cluster directly because it isn't a valid identifier, but it matches all clusters
// whose names start with `prd-2305`.
func findClusters(collection *cmv1.ClustersClient, key string) (clusters []*cmv1.Cluster, err error) {
        // Here you can hide the logic to try to find by id and then by name.
}
```

If you do that, or something similar, this code will become simpler.